src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -349,261 +179,78 @@ public virtual void GetPolicies()
         // Options can be referenced here since all JsonPropertyInfos originate from a JsonClassInfo that is cached on JsonSerializerOptions.
         protected JsonSerializerOptions Options { get; set; } = null!; // initialized in Init method
 
-        protected abstract void OnRead(ref ReadStack state, ref Utf8JsonReader reader);
-        protected abstract void OnReadEnumerable(ref ReadStack state, ref Utf8JsonReader reader);
-        protected abstract void OnWrite(ref WriteStackFrame current, Utf8JsonWriter writer);
-        protected virtual void OnWriteDictionary(ref WriteStackFrame current, Utf8JsonWriter writer) { }
-        protected abstract void OnWriteEnumerable(ref WriteStackFrame current, Utf8JsonWriter writer);
-
-        public Type ParentClassType { get; private set; } = null!;
-
-        public PropertyInfo? PropertyInfo { get; private set; }
-
-        public void Read(JsonTokenType tokenType, ref ReadStack state, ref Utf8JsonReader reader)
+        public bool ReadJsonAndAddExtensionProperty(object obj, ref ReadStack state, ref Utf8JsonReader reader)
         {
-            Debug.Assert(ShouldDeserialize);
+            object propValue = GetValueAsObject(obj)!;
 
-            JsonPropertyInfo? propertyInfo;
-            JsonClassInfo? elementClassInfo = ElementClassInfo;
-            if (elementClassInfo != null && (propertyInfo = elementClassInfo.PolicyProperty) != null)
+            IDictionary<string, object?>? dictionaryObject = propValue as IDictionary<string, object?>;
+            if (dictionaryObject != null)
             {
-                if (!state.Current.CollectionPropertyInitialized)
+                // Handle case where extension property is object-based.
```

Comment:
Oh...! This makes sense to me now. Tyvm for the comment :)

Extension dictionaries can be one of `string, object` and `string, JsonElement`, so that's why we need this null check. Cool.